### ISSUE TITLE
BSO clustering algorithm

### ISSUE THREAD

**Issue by rohanchopra** [thieu1995/mealpy#3]
I was going through your implementation of BSO and saw that you clustered all the solutions into equal size clusters randomly. Is there any specific reason for you doing that? 
The research paper that you linked mentions that they use K-Means clustering.

**Comment by thieunguyen5991** [comment]
First, the K-Means will be decreased performance when working with high-dimensional space, for example, 50, 100, 1000.

Second, I feel like no need to calculate K-Means to find the centers of a group. Our objective here is to find the point which nearest to global optima. Not find the best cluster in the population. 
You can see, in the updating process, individuals will move based on the center.
+ Now if your cluster contains only several close points together, it will only explore in small areas.
+ But if your cluster contains different points (even with far distance together), the exploration will be better, they will search in larger space than the above case.

Lastly, the K-Means clustering will slow down the algorithm, you can check by yourself.

Here is some Matlab code:
	https://www.researchgate.net/publication/322959536_Codes_in_MATLAB_for_Brain_Storm_Optimization
	https://www.mathworks.com/matlabcentral/fileexchange/60488-hypo-variance-brain-storm-optimization
	https://www.mathworks.com/matlabcentral/fileexchange/70471-an-understanding-course-on-bso
	https://www.mathworks.com/matlabcentral/fileexchange/60488-hypo-variance-brain-storm-optimization?focused=7129554&tab=function
	
At the end of the day, what we really need is the algorithm to find the nearest point to the global optima.

**Comment by rohanchopra** [comment]
Hi @thieu1995,

Sorry for the delayed response. Thanks for clarifying your modifications.

I made some changes to your code and implemented KMeans clustring using scikit-learn (PR     #4) to test the time difference.
I observed that kmeans clustering bumps up the time by approximately three times. I also observed that it did a much better job at finding the best fit when compared to the base algorithm.

### KMeans BSO
|   Run |    Best Fit |    Time |
|------:|------------:|--------:|
|    0 | 2.43685e+08 | 29.1564 |
|     1 | 2.42151e+08 | 30.0141 |
|     2 | 2.3188e+08  | 29.5265 |
|     3 | 2.11535e+08 | 33.4108 |
|     4 | 2.4127e+08  | 33.0545 |

### Base BSO
|   Run |    Best Fit |    Time |
|------:|------------:|--------:|
|     0 | 1.36081e+09 | 11.1927 |
|     1 | 1.70632e+09 | 11.1879 |
|     2 | 1.01199e+09 | 11.162  |
|     3 | 1.14893e+09 | 11.2896 |
|     4 | 1.38748e+09 | 11.2315 |


However your modified BSO did much better when compared to both the base and kmeans approach

### Improved BSO
|   Run |         Best Fit |    Time |
|------:|-----------------:|--------:|
|     0 | 106202           | 10.8593 |
|     1 |  16144.2         | 10.9116 |
|     2 |  39200.9         | 10.9024 |
|     3 |  26648.1         | 10.9773 |
|     4 |      3.01286e+06 | 10.9447 |

I tried a different approach to reduce time by removing the find_clusters function from the epoch loop and having it only at the end before the return statement.
This helped reduce the time taken by quite a bit, making it comparable with the base algorithm but, the fit was still much better with the KMeans approach.

### KMeans without frequent clustering
|   Run |    Best Fit |    Time |
|------:|------------:|--------:|
|  0 |     0 | 3.10609e+08 | 11.9124 |
|  1 |     1 | 2.86202e+08 | 11.7933 |
|  2 |     2 | 2.42978e+08 | 11.9162 |
|  3 |     3 | 2.20278e+08 | 12.2182 |
|  4 |     4 | 2.17304e+08 | 12.3023 |

I think the rationale is that Step 6 c) gets affected by how good the clusters are. In your implementation the points get randomly assigned to a cluster which does not help the combining points of two clusters step (Step 6 c) which happens with a high probability.

One more major change I implemented apart from using KMeans was from the original paper which said:
"In Step 6.b.i of the BSO procedure shown in Table 3, a cluster is selected with probability p6bi, which is proportional to the number of individuals in the cluster." this can be seen in line #245 of BSO.py. This modification did not change the loss or time required much.

### Results with the probability modification
|   Run |    Best Fit |    Time |
|------:|------------:|--------:|
|     0 | 2.56397e+08 | 31.22   |
|     1 | 2.39896e+08 | 33.9796 |
|     2 | 2.59535e+08 | 33.8842 |
|     3 | 2.46954e+08 | 33.9066 |
|     4 | 2.19942e+08 | 33.5345 |

Do let me know your thoughts.

Thanks
Rohan

**Comment by nour-elhuda** [comment]
I need help in solving a task scheduling in cloud computing using pso-cso by use the output from the pso to be input for cso for the same iteration does that work? Any suggestion please plesase

**Comment by thieu1995** [comment]
@rohanchopra Hi. Sorry for the late reply.

From my limited knowledge and experiment. There is no technique which can enhance Metaheuristic Algorithms better than Levy-flight. You can put it in every MHAs, no down it will become better. That is the reason why they are producing tons of papers out there just by doing that. (Of course, I didn't count some improved techniques from Gradient Descent in Neural Network like Adam, Adagrad,... because they are from a different perspective).

So for me, MHAs should keep its characteristics like easy to implement, simple to understand, efficient for some specific problems... Therefore, if you are using KMeans or any kind of clustering-based algorithms it still can easily implement (by calling from another library) but it makes the algorithm is more complicated. Anyway, you did really good work there.

**Comment by thieu1995** [comment]
Hi @nour-elhuda ,

Please don't spam on my channels (GitHub, email, Facebook,...). As we talked previously in FB, you shouldn't do that (using the output of PSO to be input for CSO in the same iterations). I also gave you some suggestions and documents to study other ways. 
Currently, I'm doing a project which solving task scheduling in cloud computing but with multiple/many objectives instead of single (one) objective. The code is done but currently not available for users yet. I will just leave the link [here](https://github.com/thieu1995/IFCB), you can access when its come publicity. 
Also, as the first notification of this version of this library. I'm working on a new open-source library for solving multiple/many objectives using Metaheuristic Algorithms (benchmark functions). You can check it [here.](https://github.com/thieu1995/momapy)
